doc/ref_arch/kubernetes/chapters/chapter04.md

```diff
@@ -18,26 +18,45 @@
+identifiers, which will follow the pattern: `ra2.<section>.<index>`, e.g.
+`ra2.ch.001` for the first requirement in the Container Host section.  These
+specifications will then be used as requirements input for the Kubernetes
+Reference Implementation and any Vendor or Community Implementations.
+
+Figure 4-1 below shows the architectural components that are described in the
+subsequent sections of this chapter.
+
+<p align="center"><img src="../figures/ch04_k8s_architecture.png"
+alt="Kubernetes Reference Architecture" Title="Kubernetes Reference
+Architecture" width="65%"/></p>
 <p align="center"><b>Figure 4-1:</b> Kubernetes Reference Architecture</p>
 
 <a name="4.2"></a>
 ## 4.2 Container Host
 
-This section describes the configuration that will be applied to the physical or virtual machine and an installed Operating System.
-In order for a Container Host to be conformant with the Reference Architecture it must be implemented as per the following specifications:
+This section describes the configuration that will be applied to the physical or
+virtual machine and an installed Operating System. In order for a Container Host
+to be conformant with the Reference Architecture it must be implemented as per
+the following specifications:
 
 |Ref|Specification|Details|Requirement Trace|
 |---|---|---|---|
-|`ra2.ch.001`||||
-|`ra2.ch.002`||||
-|`ra2.ch.003`||||
+|`ra2.ch.001`|Huge Pages|Huge Pages must be enabled within the Container Host OS, with at least one each of 2048KiB and 1048576KiB page sizes, exposing schedulable resources `hugepages-2Mi` and `hugepages-1Gi`.|[infra.com.cfg.004](./chapter02.md#223-cloud-infrastructure-software-profile-requirements)|
+|`ra2.ch.002`|SR-IOV capable NICs|The physical machine on which the Container Host runs must be equipped with NICs that are SR-IOV capable (e.g. [this list from Intel](../../../../../../../../intel/sriov-network-device-plugin#supported-sr-iov-nics)).|[e.cap.013](./chapter02.md#223-cloud-infrastructure-software-profile-requirements)|
+|`ra2.ch.003`|SR-IOV Virtual Functions|SR-IOV virtual functions (VFs) must be configured within the Container Host OS, as the SR-IOV Device Plugin does not manage the creation of these VFs. (e.g. [vf-setup.md](../../../../../../../../intel/sriov-network-device-plugin/blob/master/docs/vf-setup.md)).|[e.cap.013](./chapter02.md#223-cloud-infrastructure-software-profile-requirements)|
+|`ra2.ch.004`|CPU Simultaneous Multi-Threading (SMT)|SMT must be enabled in the BIOS on the physical machine on which the Container Host runs.|[infra.hw.cpu.cfg.004](./chapter02.md#224-cloud-infrastructure-hardware-profile-requirements)|
+|`ra2.ch.005`|CPU Allocation Ratio|For Kubernetes nodes running as Virtual Machines, ensure the CPU allocation ratio between vCPU and physical CPU is 1:1 (note that it is not possible for the kubelet to over-allocate CPU resources.)|[infra.com.cfg.001](./chapter02.md#223-cloud-infrastructure-software-profile-requirements)|
```

Comment:
1: I guess you wanted to say thread, not core. In case HT is enabled 1 vCPU is 1 thread usually
2: not just that it is possible, it is the default setting :) deploy any Pods without limits, start up some infinite loops and watch them eat the resources
requests cannot be overallocated sure, but limits very much can be more than the available physical CPU shares

Reply:
Interesting, I was under the impression that the "Allocatable" was a hard limit, regardless of the requests made by Pods (i.e. allocatable would get used, next Pod would be stuck in pending state?)

Reply:
only if your Pod state "requests". if you don't state a request you will be scheduled

Reply:
from a live system:
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource                        Requests     Limits
  --------                        --------     ------
  cpu                             1628m (3%)   7498m (17%)

The help text itself says limits might be overcommited :)

Reply:
thanks - will tweak the wording

Reply:
OK I think I understand now - please correct any mistakes!  
- Sum(`limits`) > `allocatable` is valid.
- Sum(`requests`) > `allocatable` is not valid.
- If you do specify `limits` but don't specify `requests` then `requests`==`limits` and over-commitment isn't possible (because sum(`requests`) must be < `allocatable`).
- If you do specify `requests` < `limits` then you are in the shared CPU pool, and over-commitment (i.e. sum(`limits`) > allocatable) is possible as long as sum(`requests`) < `allocatable`.

Is that right?

Reply:
not sure if request = limit is added to a Pod wrt CPU, but otherwise correct, yes

Reply:
The last example [here](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/#static-policy) suggests:

> This pod runs in the Guaranteed QoS class because only limits are specified and requests are set equal to limits when not explicitly specified.

Reply:
Levente, does this new wording work for you? https://github.com/cntt-n/CNTT/pull/1805/commits/a2904850d7710add072f05421b377c069eac190e

Reply:
so yeah it is factually correct now. remaining question is semantic based :)
if we truly want to control overcommitment ratio, I think a similar sentence for the summarized limits might be also required
i.e. requests and limits both must be defined for all containers, and summa requests must be lower than allocatable, and summa limits must be lower than 2Xallocatable -> guaranteed 1:2 overcommitment ratio

if this RQ is about controlling overcommitment ratio then limits should be taken into account too IMO; if not then nevermind :)

Reply:
> i.e. requests and limits both must be defined for all containers, and summa requests must be lower than allocatable, and summa limits must be lower than 2Xallocatable -> guaranteed 1:2 overcommitment ratio

Except that the RM states that allocation ratio must be 1:1 for all profiles.  I'll update to reflect what you've mentioned about limits - agreed.

Reply:
yeah then the same, but summa limits must equal summa requests

Reply:
OK, updated in https://github.com/cntt-n/CNTT/pull/1805/commits/13b0d5834bea31bd72b5951116734dd273dbdb58